[MRG] ENH add Poisson splitting criterion for single trees

lorentzenchr opened the pull request:
#### Reference Issues/PRs
Contributes to #16668.

#### What does this implement/fix? Explain your changes.
This PR adds a Poisson splitting criterion to `DecisionTreeRegressor` and `ExtraTreeRegressor` with option `criterion="poisson"`.

#### Any other comments?
The implemented splitting criterion forbids splits with child nodes having `sum(y_i) = 0` as non-positive predictions are not allowed for Poisson regression/loss. This is the simplest solution to this problem, i.e. without introducing another option.

If this is good to go and merged, the additional option `"poisson"` can be propagated to `RandomForestRegressor` and other tree based ensemble regressors.

Commit:
ENH add poisson splitting criterion for trees

Commit:
TST add tests for poisson splitting in trees

Commit:
BUG fix and test fobidden zero node split for Poisson

Commit:
TST include poisson criterion in test_diabetes_underfit

NicolasHug commented on sklearn/tree/_criterion.pyx:
```diff
@@ -26,11 +26,17 @@ import numpy as np
 from ._utils cimport WeightedMedianCalculator
 
+
+cdef double INFINITY = np.inf
```
You can use `from numpy.math cimport INFINITY` instead

NicolasHug commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -289,9 +298,9 @@ def test_diabetes_underfit(name, Tree, criterion, max_depth):
-    score = mean_squared_error(diabetes.target, reg.predict(diabetes.data))
-    assert score < 60 and score > 0, (
-        f"Failed with {name}, criterion = {criterion} and score = {score}"
+    loss = mean_squared_error(diabetes.target, reg.predict(diabetes.data))
```
does it make sense to check the mse when the tree was trained with a poisson criteria? The need for a limit of 4000 seems to indicate that the check is super loose?

NicolasHug replied:
also nit but score seemed like the correct term here. There's no loss in these trees

lorentzenchr replied:
MSE elicits the expectation, so does Poisson deviance. So yes, it makes sense.
I tried to use mean_poisson_deviance, but that did not change much, i.e. still needs a high `max_loss`. I don't know why.

"score" in scikit-learn means "higher values are better". Error and loss, for me, are often synonyms. I can change into "error"?

lorentzenchr replied:
I'm also skeptical here. I changed to use Poisson deviance as score for `"poisson"` splitter. It's much better now.

NicolasHug commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -1948,3 +1964,24 @@ def check_apply_path_readonly(name):
+
+
+@pytest.mark.parametrize("criterion", ["mse", "poisson"])
+@pytest.mark.parametrize("tree", REG_TREES.values())
```
```suggestion
@pytest.mark.parametrize("Tree", REG_TREES.values())
```

NicolasHug commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -1948,3 +1964,24 @@ def check_apply_path_readonly(name):
+@pytest.mark.parametrize("criterion", ["mse", "poisson"])
+@pytest.mark.parametrize("tree", REG_TREES.values())
+def test_balance_property(criterion, tree):
+    """Test that sum(y_pred)=sum(y_true) on training set."""
```
maybe add a comment that this can't be true for mae because the median is used instead of the mean (if that's true, I haven't double checked)?
what about friedman_mse?

lorentzenchr replied:
I added `"friedman_mse"` - I had not double checked - and a comment about mae and median.
This could also be tested for each leaf.

NicolasHug commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -1948,3 +1964,24 @@ def check_apply_path_readonly(name):
+
+
+def test_poisson_zero_nodes():
+    """Test that sum(y)=0 and therefore y_pred=0 never happens on nodes."""
```
```suggestion
    """Test that sum(y)=0 and therefore y_pred=0 is forbidden on nodes."""
```

I was surprised this wasn't actually happening considering the dataset and then I saw your comment about this being enforced

NicolasHug replied:
also nit but our convention for tests is to use comments instead of docstrings (don't ask me why ;) )

NicolasHug commented on sklearn/tree/_criterion.pyx:
```diff
@@ -1325,3 +1327,122 @@ cdef class FriedmanMSE(MSE):
+
+    Poisson deviance = 2/n * sum(y_true * log(y_true/y_pred) + y_pred - y_true)
+
+    Note that the deviance is >= 0, and for trees with `y_pred = mean(y_true)`,
```
```suggestion
    Note that the deviance is >= 0, and since we have `y_pred = mean(y_true)` at the leaves,
```

NicolasHug commented on sklearn/tree/_criterion.pyx:
```diff
@@ -1325,3 +1327,122 @@ cdef class FriedmanMSE(MSE):
+    Note that the deviance is >= 0, and for trees with `y_pred = mean(y_true)`,
+    one always has `sum(y_pred - y_true) = 0`. It remains the implemented
+    impurity:
+        1/n * sum(y_i * log(y_i/y_pred)
```
```suggestion
        1/n * sum(y_true * log(y_true/y_pred)
```

NicolasHug commented on sklearn/tree/_criterion.pyx:
```diff
@@ -1325,3 +1327,122 @@ cdef class FriedmanMSE(MSE):
+                if self.sample_weight != NULL:
+                    w = self.sample_weight[i]
+
+                impurity += w * xlogy(self.y[i, k], self.y[i, k]/y_mean)
```
```suggestion
                impurity += w * xlogy(self.y[i, k], self.y[i, k] / y_mean)
```

NicolasHug commented on sklearn/tree/_criterion.pyx:
```diff
@@ -1325,3 +1327,122 @@ cdef class FriedmanMSE(MSE):
+    impurity:
+        1/n * sum(y_i * log(y_i/y_pred)
+    """
+    # FIXME in 0.25:
```
Is this relevant considering we have `proxy_impurity_improvement`?

lorentzenchr replied:
Isn't the impurity calculated for each node (after a split)? If so, the FIXME might improve performance as `xlogy` must be calculated for every sample in the node. There is no algebraic simplification.
I have not done benchmakrs though.

NicolasHug commented on sklearn/tree/_criterion.pyx:
```diff
@@ -1325,3 +1327,122 @@ cdef class FriedmanMSE(MSE):
+                if self.sample_weight != NULL:
+                    w = self.sample_weight[i]
+
+                impurity_left[0] += w * xlogy(y[i, k], y[i, k]/y_mean)
```
```suggestion
                impurity_left[0] += w * xlogy(y[i, k], y[i, k] / y_mean)
```

NicolasHug commented on sklearn/tree/_criterion.pyx:
```diff
@@ -1325,3 +1327,122 @@ cdef class FriedmanMSE(MSE):
+                if self.sample_weight != NULL:
+                    w = self.sample_weight[i]
+
+                impurity_right[0] += w * xlogy(y[i, k], y[i, k]/y_mean)
```
```suggestion
                impurity_right[0] += w * xlogy(y[i, k], y[i, k] / y_mean)
```

NicolasHug commented on sklearn/tree/_criterion.pyx:
```diff
@@ -1325,3 +1327,122 @@ cdef class FriedmanMSE(MSE):
+        cdef DOUBLE_t w = 1.0
+
+        impurity_left[0] = 0.
+        for k in range(self.n_outputs):
```
any chance we can collect some statistics only once for the left node, and then get it for free for the right node as done for MSE? (Instead of looping twice?)

lorentzenchr replied:
See the `FIXME` comments. Omitting constant terms, it would suffice to compute `y_mean`. The full Poisson deviance needs `y_true_i * log(y_true_i)` for every sample `i`. Maybe, one could define this as as statistics like/instead of `sq_sum_total`? 

Since I wanted a small PR, I did not consider this - and also, I was a bit frightened by the implementation of `cdef class MAE`:smirk:

NicolasHug reviewed (commented):
thanks @lorentzenchr , made a first pass, looks good

we'll also need some docs to the UG ;)

Commit:
address review comments

lorentzenchr commented:
> we'll also need some docs to the UG ;)

Might be worth considering 😄

lorentzenchr commented:
The user guide does not mention `"friedman_mse"`. Where should `"poisson"` be mentioned? Under https://scikit-learn.org/stable/modules/tree.html#regression-criteria?

NicolasHug commented:
Yes under  scikit-learn.org/stable/modules/tree.html#regression-criteria for the formula. (Feel free to add Friedman if you feel like it ;) )

Also a small paragraph in https://scikit-learn.org/stable/modules/tree.html#regression to distinguish criteria would be helpful. E.g. mae is less sensitive to outliers, friedman mse is (?????), and poisson is (?????)

Commit:
DOC add Poisson deviance as impurity in user guide

Commit:
DOC improve tree section of user guide

lorentzenchr commented:
@NicolasHug I added the Poisson deviance to the user guide and also did some more improvements. Please tell me if you like them.

About the `"friedman_mse"` - maybe better as separate issue: It is never mentioned in the User Guide or examples. After consulting the literature [1], I don't understand why it is implemented/API-exposed for `DecisionTreeRegressor`. It seems to have its place in multiclass gradient-boosted trees.

[1] https://statweb.stanford.edu/~jhf/ftp/trebst.pdf

Commit:
DOC miiiiiinor text improvement

Commit:
Merge branch 'master' into poisson_tree

lorentzenchr commented:
@NicolasHug I merged master to resolve merge conflicts. I think I addressed all the comments from your first review pass. Anything else I can do?

thomasjpfan commented on sklearn/tree/_classes.py:
```diff
@@ -161,6 +164,14 @@ def fit(self, X, y, sample_weight=None, check_input=True,
                     raise ValueError("No support for np.int64 index based "
                                      "sparse matrices")
 
+            if self.criterion in ["poisson", _criterion.Poisson]:
```
I would rather be more restrictive just in case.

```suggestion
            if self.criterion == "poisson":
```

thomasjpfan commented on sklearn/tree/_criterion.pyx:
```diff
@@ -26,11 +26,15 @@ import numpy as np
 np.import_array()
 
+from numpy.math cimport INFINITY
+from scipy.special.cython_special cimport xlogy
```
Wow this is cool

thomasjpfan commented on sklearn/tree/_criterion.pyx:
```diff
@@ -1325,3 +1325,122 @@ cdef class FriedmanMSE(MSE):
+            y_mean_right = self.sum_right[k] / self.weighted_n_right
+            if (y_mean_left <= 0) or (y_mean_right <= 0):
+                # Poisson loss does not allow non-positive predictions. We
+                # therefore forbid splits that have child nodes withs
```
```suggestion
                # therefore forbid splits that have child nodes with
```

thomasjpfan commented on sklearn/tree/_criterion.pyx:
```diff
@@ -1325,3 +1325,122 @@ cdef class FriedmanMSE(MSE):
+        for k in range(self.n_outputs):
+            y_mean_left = self.sum_left[k] / self.weighted_n_left
+            y_mean_right = self.sum_right[k] / self.weighted_n_right
+            if (y_mean_left <= 0) or (y_mean_right <= 0):
```
We can check dividing by the weights?

lorentzenchr replied:
What do you mean?

thomasjpfan replied:
I forgot a word:

> We can do this check **before** dividing by the weights.

thomasjpfan reviewed (commented):
Thank you for the PR @lorentzenchr 

This is a partial review.

Commit:
address review comments

Commit:
Check positive child sums before calculating means

lorentzenchr commented:
@NicolasHug @thomasjpfan I hope you're still able to see the forest for the trees, in particular for boosted and histrogrammed trees.:smirk: Just a little, well-meant reminder for this PR.

NicolasHug commented:
I tried to push an empty commit but it looks like I'm not allowed to push to this PR. Could you just merge with master @lorentzenchr so that the docs get rendered?

NicolasHug commented on sklearn/tree/_criterion.pyx:
```diff
@@ -1325,3 +1325,122 @@ cdef class FriedmanMSE(MSE):
+                impurity_left[0] += w * xlogy(y[i, k], y[i, k] / y_mean)
+        impurity_left[0] /= self.weighted_n_left * self.n_outputs
+
+        impurity_right[0] = 0.
+        for k in range(self.n_outputs):
+            y_mean = self.sum_right[k] / self.weighted_n_right
+            for p in range(pos, end):
+                i = self.samples[p]
+
+                if self.sample_weight != NULL:
+                    w = self.sample_weight[i]
+
+                impurity_right[0] += w * xlogy(y[i, k], y[i, k] / y_mean)
+        impurity_right[0] /= self.weighted_n_right * self.n_outputs
```
this bit of code seems to be duplicated 3 times. Would it make sense to extract it into a private method so that we could use it here and in `node_impurity`?

lorentzenchr replied:
Same could be done for MAE.

NicolasHug reviewed (commented):
thanks for the patience @lorentzenchr , made another pass

It would be nice to have some high level tests ensuring that using Poisson results in a better score than using e.g. MSE on a specific problem where Poisson is expected to perform better?

I also feel like the UG could be a bit more detailed on which cases one might want to use the Poisson criteria vs MSE/MAE.

Also could you quickly benchmark the Poisson criteria vs MSE? I'm curious about the "overhead" incurred in `children_impurity`

lorentzenchr committed:
trigger doc build

lorentzenchr committed:
Merge branch 'master' into poisson_tree

lorentzenchr commented on doc/modules/tree.rst:
```diff
@@ -468,37 +473,40 @@ observations, let
     H(X_m) = 1 - \max(p_{mk})
 
-where :math:`X_m` is the training data in node :math:`m`
+Here, :math:`X_m` is the training data in node :math:`m`.
```
After some thought (and trial), I'd prefer to open an issue to improve the UG of trees:
- Improve mathematical formulation, particularly the regions, i.e. unify Q, R, X.
- Mention what a prediction per node is.
- Consistent shape parameters and more.

lorentzenchr replied:
I gave the formulae a trial. What do you think?

lorentzenchr commented on doc/modules/tree.rst:
```diff
@@ -508,6 +516,15 @@ Mean Squared Error:
+
+.. math::
+
+    \bar{y}_m = \frac{1}{N_m} \sum_{i \in N_m} y_i
```
I tried that one, but again, it doesn't make the formulae better or less wrong. Separate issue/PR?

NicolasHug replied:
it's more about consistency than it is about correctness I think. Do we really need to use R_m? It seems we could simply rely on X_m and use N_m purely as a number and not as a set of indices.

NicolasHug replied:
But yes I'm happy to have this in another PR. It's just that it might delay this one a bit more. So if not too many changes are involved, it seems that doing it here would be more efficient, but you probably have a better overview so it's your call ;)

lorentzenchr commented on sklearn/tree/_criterion.pyx:
```diff
@@ -279,8 +275,10 @@ cdef class ClassificationCriterion(Criterion):
-        children samples[start:start] and samples[start:end].
+        """Initialize the criterion.
+
+        This initializes the criterion at node samples[start:end] and children
+        samples[start:start] and samples[start:end].
```
I did not write that part, just formatted the docstring which shows your point above about cosmetic changes.

lorentzenchr commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -1950,6 +1966,30 @@ def test_apply_path_readonly_all_trees(name):
 
+@pytest.mark.parametrize("criterion", ["mse", "friedman_mse", "poisson"])
+@pytest.mark.parametrize("Tree", REG_TREES.values())
+def test_balance_property(criterion, Tree):
```
Thanks. An often neglected, but very important property. I'm wondering if some common test would make sense - maybe not.

NicolasHug replied:
you a common test for which other estimators?

I think we could definitely have this for the histogram-GBDT trees but I can't think of any other one

lorentzenchr replied:
It would be for a lot of models using MSE, Tweedie deviance or Log Loss. Counter examples are linear models without intercept and uncentered data.

lorentzenchr committed:
address review comments

lorentzenchr committed:
TST compare poisson split vs mse

lorentzenchr commented on sklearn/tree/_criterion.pyx:
```diff
@@ -1325,3 +1325,122 @@ cdef class FriedmanMSE(MSE):
+            else:
+                y_mean_left = self.sum_left[k] / self.weighted_n_left
+                y_mean_right = self.sum_right[k] / self.weighted_n_right
+                proxy_impurity_left += y_mean_left * log(y_mean_left)
```
Oooops. I think you are right:blush:

lorentzenchr commented:
@NicolasHug Thanks for your last/nest review round.
- I decided to keep the FIXME comment and elaborate a bit more.
- Added test poisson vs mse, so far failing and no clue.
- So far, I'd like to focus on implementation and not on User Guide (maybe separate issue).
- [ ] Refer to GLM User Guide about when to use `"poisson"`

lorentzenchr committed:
MNT reduce redundant code with helper poisson_loss

lorentzenchr committed:
TST make test_poisson_vs_mse pass

lorentzenchr committed:
TST improve test_diabetes_underfit

lorentzenchr committed:
DOC consistent formulae

lorentzenchr committed:
Merge branch 'master' into poisson_tree

lorentzenchr committed:
DOC add hint as to when to use Poisson criterion

lorentzenchr committed:
DOX fix fomula super- and subscripts

lorentzenchr commented:
@NicolasHug @thomasjpfan Do you have time to finish your reviews? With this PR, once it is correct, sound and safe, it's only a matter of propagating options to have Poisson splitting criterion in random forests which I would really like to see.

NicolasHug commented on doc/modules/tree.rst:
```diff
@@ -429,99 +433,111 @@ Mathematical formulation
 
-Let the data at node :math:`m` be represented by :math:`Q`. For
-each candidate split :math:`\theta = (j, t_m)` consisting of a
+Let the data at node :math:`m` be represented by :math:`Q_m` with :math:`N_m`
```
Thanks for updating these. It looks better now.

NicolasHug commented on doc/modules/tree.rst:
```diff
@@ -429,99 +433,111 @@ Mathematical formulation
 .. math::
 
-    median(y)_m = \underset{i \in N_m}{\mathrm{median}}(y_i)
+    \bar{y}_m = \frac{1}{N_m} \sum_{y \in Q_m} y
```
IMHO we don't need to define it twice. It's defined just above already

NicolasHug commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -277,10 +290,11 @@ def test_diabetes_overfit(name, Tree, criterion):
-    [("mse", 15), ("mae", 20), ("friedman_mse", 15)]
+    "criterion, max_depth, max_loss",
+    [("mse", 15, 60), ("mae", 20, 60), ("friedman_mse", 15, 60),
+     ("poisson", 15, 30)]
```
I remember `max_loss` to be very high previously. Did you fix something?

lorentzenchr replied:
For `test_diabetes_underfit`, I changed the metric to poisson deviance for `criterion='poisson'`, see [this commit](https://github.com/scikit-learn/scikit-learn/pull/17386/commits/a5d84bbff957ec490370c70e0ff4cb185fa14fd9).

NicolasHug commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -289,9 +303,13 @@ def test_diabetes_underfit(name, Tree, criterion, max_depth):
+    else:
+        loss = mean_squared_error(diabetes.target, reg.predict(diabetes.data))
+    assert 0 < loss < max_loss, (
+        f"Failed with {name}, criterion = {criterion} and loss = {loss}"
```
IMHO the error message doesn't help much more than what `pytest` would already tell us. I understand it was there before, but we can remove it

NicolasHug commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -1950,6 +1966,30 @@ def test_apply_path_readonly_all_trees(name):
+    y = [0, 0, 0, 0, 1, 2, 3, 4]
+    reg = DecisionTreeRegressor(criterion="poisson", random_state=1)
+    reg.fit(X, y)
+    assert np.all(reg.predict(X) > 0)
```
ping on this one ;)

lorentzenchr replied:
I added a further dataset where **this test fails**:bomb:. As the impurity improvement returns -INFINITY (same as initialized value), it should not allow those splits. I'm investigating the splitter code where this might happen.
Interestingly, the impurity value for those nodes that predict zero is `np.nan`.

thomasjpfan replied:
> Interestingly, the impurity value for those nodes that predict zero is np.nan.

Is this issue specific to this PR or an general issue with the tree splitting?

lorentzenchr replied:
The `np.nan`was a consequence of a split for Poisson that was not allowed with only y=0 in the node. This led to `y_pred=0` which produced Poisson loss = np.nan.
This is all gone with commit 82e35867c2aa01e3c2f944c100564a799dc01655 "BUG check for y_mean <= 0 in poisson_loss".

NicolasHug commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -1950,6 +1975,74 @@ def test_apply_path_readonly_all_trees(name):
+def test_poisson_vs_mse():
+    # For Poisson distributed target, Poisson loss should give better results
+    # than least squares measured in Poisson deviance as metric.
+    # Compare to
+    # sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py
```
what do you mean by this? Does this mean "we have a similar test in ..." ?

NicolasHug commented on sklearn/tree/_criterion.pyx:
```diff
@@ -1330,3 +1330,117 @@ cdef class FriedmanMSE(MSE):
+                                      SIZE_t end,
+                                      DOUBLE_t* y_sum,
+                                      DOUBLE_t weight_sum) nogil:
+        """Helper function to compute Poisson loss (~deviance)."""
```
```suggestion
        """Helper function to compute Poisson loss (~deviance) of a given node."""
```

NicolasHug commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -1950,6 +1975,74 @@ def test_apply_path_readonly_all_trees(name):
+    # than least squares measured in Poisson deviance as metric.
+    # Compare to
+    # sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py
+    # Note: Some fine tuning was needed to have metric_poi < metric_dummy on
+    # the test set!
```
Can you confirm that this was an overfitting problem, i.e. that any reasonable value > 10 for `min_samples_split` would still pass the test? Otherwise there might be some underlying issue

lorentzenchr replied:
For `min_sample_split>=45` MSE gives better poisson deviance on the test set, see details.
The `n_test` is the most fine-tuned one. For `n_test=100`, for instance, the dummy regressor gives better poisson deviances on the test set. Given that test deviance is 4 times higher than train deviance, this is an overfitting issue. Any suggestions?

<details>

```python
import numpy as np
from sklearn.datasets import make_low_rank_matrix
from sklearn.dummy import DummyRegressor
from sklearn.metrics import mean_poisson_deviance, mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor


rng = np.random.RandomState(42)
n_train, n_test, n_features = 500, 500, 10
X = make_low_rank_matrix(n_samples=n_train + n_test,
                                  n_features=n_features, random_state=rng)
# We create a log-linear Poisson model and downscale coef as it will get
# exponentiated.
coef = rng.uniform(low=-2, high=2, size=n_features) / np.max(X, axis=0)
y = rng.poisson(lam=np.exp(X @ coef))
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=n_test,
                                                    random_state=rng)

for msp in range(2, 81):
    tree_poi = DecisionTreeRegressor(criterion="poisson",
                                     min_samples_split=msp,
                                     random_state=rng)
    tree_mse = DecisionTreeRegressor(criterion="mse",
                                     min_samples_split=msp,
                                     random_state=rng)
    tree_poi.fit(X_train, y_train)
    tree_mse.fit(X_train, y_train)
    a = mean_poisson_deviance(y_test, tree_poi.predict(X_test))
    b = mean_poisson_deviance(y_test, np.clip(tree_mse.predict(X_test), 1e-15, None))
    print(f"min_sample_split={msp:2d}  Test Deviance  poisson={a:3.2f},  mse={b: 4.2f}")
```
Output (for splitting criterion poisson and mse)
```
min_sample_split= 2  Test Deviance  poisson=5.27,  mse=21.41
min_sample_split= 3  Test Deviance  poisson=5.08,  mse=21.10
min_sample_split= 4  Test Deviance  poisson=5.15,  mse=22.67
min_sample_split= 5  Test Deviance  poisson=5.43,  mse=19.81
min_sample_split= 6  Test Deviance  poisson=4.94,  mse=17.21
min_sample_split= 7  Test Deviance  poisson=4.73,  mse=16.62
min_sample_split= 8  Test Deviance  poisson=5.30,  mse=15.79
min_sample_split= 9  Test Deviance  poisson=4.89,  mse=14.02
min_sample_split=10  Test Deviance  poisson=4.85,  mse=12.14
min_sample_split=11  Test Deviance  poisson=4.89,  mse=11.24
min_sample_split=12  Test Deviance  poisson=4.77,  mse=10.68
min_sample_split=13  Test Deviance  poisson=4.70,  mse=8.57
min_sample_split=14  Test Deviance  poisson=4.72,  mse=8.50
min_sample_split=15  Test Deviance  poisson=4.66,  mse=8.53
min_sample_split=16  Test Deviance  poisson=4.62,  mse=8.45
min_sample_split=17  Test Deviance  poisson=4.52,  mse=7.42
min_sample_split=18  Test Deviance  poisson=4.43,  mse=7.38
min_sample_split=19  Test Deviance  poisson=4.40,  mse=7.38
min_sample_split=20  Test Deviance  poisson=4.28,  mse=7.42
min_sample_split=21  Test Deviance  poisson=4.27,  mse=7.18
min_sample_split=22  Test Deviance  poisson=4.28,  mse=6.99
min_sample_split=23  Test Deviance  poisson=4.32,  mse=7.05
min_sample_split=24  Test Deviance  poisson=4.32,  mse=7.01
min_sample_split=25  Test Deviance  poisson=4.32,  mse=6.95
min_sample_split=26  Test Deviance  poisson=4.32,  mse=6.99
min_sample_split=27  Test Deviance  poisson=4.31,  mse=7.01
min_sample_split=28  Test Deviance  poisson=4.31,  mse=6.98
min_sample_split=29  Test Deviance  poisson=4.31,  mse=6.94
min_sample_split=30  Test Deviance  poisson=4.31,  mse=6.94
min_sample_split=31  Test Deviance  poisson=4.31,  mse=6.96
min_sample_split=32  Test Deviance  poisson=4.31,  mse=7.00
min_sample_split=33  Test Deviance  poisson=4.31,  mse=7.00
min_sample_split=34  Test Deviance  poisson=4.31,  mse=6.96
min_sample_split=35  Test Deviance  poisson=4.31,  mse=6.96
min_sample_split=36  Test Deviance  poisson=4.29,  mse=6.90
min_sample_split=37  Test Deviance  poisson=4.29,  mse=6.90
min_sample_split=38  Test Deviance  poisson=4.29,  mse=4.53
min_sample_split=39  Test Deviance  poisson=4.28,  mse=4.47
min_sample_split=40  Test Deviance  poisson=4.28,  mse=4.47
min_sample_split=41  Test Deviance  poisson=4.29,  mse=4.47
min_sample_split=42  Test Deviance  poisson=4.28,  mse=4.47
min_sample_split=43  Test Deviance  poisson=4.28,  mse=4.44
min_sample_split=44  Test Deviance  poisson=4.29,  mse=4.87
min_sample_split=45  Test Deviance  poisson=4.28,  mse=4.00
min_sample_split=46  Test Deviance  poisson=4.29,  mse=4.03
min_sample_split=47  Test Deviance  poisson=4.28,  mse=4.03
min_sample_split=48  Test Deviance  poisson=4.29,  mse=3.97
min_sample_split=49  Test Deviance  poisson=4.29,  mse=3.99
min_sample_split=50  Test Deviance  poisson=4.29,  mse=3.99
min_sample_split=51  Test Deviance  poisson=4.27,  mse=4.01
min_sample_split=52  Test Deviance  poisson=4.28,  mse=4.01
min_sample_split=53  Test Deviance  poisson=4.27,  mse=4.01
min_sample_split=54  Test Deviance  poisson=4.27,  mse=4.01
min_sample_split=55  Test Deviance  poisson=4.27,  mse=4.01
min_sample_split=56  Test Deviance  poisson=4.28,  mse=3.99
min_sample_split=57  Test Deviance  poisson=4.28,  mse=3.99
min_sample_split=58  Test Deviance  poisson=4.28,  mse=3.99
min_sample_split=59  Test Deviance  poisson=4.28,  mse=4.02
min_sample_split=60  Test Deviance  poisson=4.27,  mse=4.07
min_sample_split=61  Test Deviance  poisson=4.27,  mse=4.07
min_sample_split=62  Test Deviance  poisson=4.28,  mse=4.06
min_sample_split=63  Test Deviance  poisson=4.27,  mse=4.06
min_sample_split=64  Test Deviance  poisson=4.28,  mse=3.91
min_sample_split=65  Test Deviance  poisson=4.28,  mse=3.91
min_sample_split=66  Test Deviance  poisson=4.27,  mse=3.91
min_sample_split=67  Test Deviance  poisson=4.27,  mse=3.91
min_sample_split=68  Test Deviance  poisson=4.27,  mse=3.91
min_sample_split=69  Test Deviance  poisson=4.27,  mse=3.91
min_sample_split=70  Test Deviance  poisson=4.27,  mse=3.91
min_sample_split=71  Test Deviance  poisson=4.27,  mse=3.92
min_sample_split=72  Test Deviance  poisson=4.26,  mse=3.92
min_sample_split=73  Test Deviance  poisson=4.26,  mse=3.92
min_sample_split=74  Test Deviance  poisson=4.26,  mse=3.92
min_sample_split=75  Test Deviance  poisson=4.26,  mse=3.92
min_sample_split=76  Test Deviance  poisson=4.26,  mse=3.92
min_sample_split=77  Test Deviance  poisson=4.26,  mse=3.92
min_sample_split=78  Test Deviance  poisson=4.26,  mse=3.92
min_sample_split=79  Test Deviance  poisson=4.26,  mse=3.92
min_sample_split=80  Test Deviance  poisson=4.26,  mse=3.92
```
</details>

NicolasHug reviewed (approved):
Thanks @lorentzenchr , made some last comments but LGTM.

I'd still be curious about some quick benchmarking in terms of `fit` time w.r.t. to MSE ;)

lorentzenchr committed:
address review comments

lorentzenchr commented:
And here come some benchmarks.
![image](https://user-images.githubusercontent.com/15324633/95012609-517dc980-063a-11eb-94c2-fe84398ad641.png)
![image](https://user-images.githubusercontent.com/15324633/95012629-665a5d00-063a-11eb-818b-0fad7d5bee80.png)
I was expecting a slow down of poisson compared to mse between 5 and 10, not up to 30+.

<details>

```python
from time import time

import matplotlib.pyplot as plt
import numpy as np
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.datasets import make_regression
from sklearn.tree import DecisionTreeRegressor

from neurtu import delayed, timeit, Benchmark


def make_criterion_benchmark(
    n_features=50,
    ax=None,
    n_samples_range=np.linspace(100, 10_000, num=10, dtype=np.int),
    task="regression",
    repeat=1,
    metric_name="wall_time",
    **kwargs,
):
    """Benchmark DecisionTree for criterion 'mse' and 'poisson'"""

    dataset_kwargs = dict(
        n_samples=n_samples_range.max(),
        n_features=n_features,
        n_informative=n_features * 2 // 3,
    )
    X, y = make_regression(
        effective_rank=n_features * 2 // 3, tail_strength=0.6, **dataset_kwargs
        )

    # some excess zeros
    y[(-1 < y) & (y < 0)] = 0
    # make sure the target is positive
    y = np.abs(y)

    bench = []

    criteria = ['mse', 'poisson']
    
    # nicer numbers
    if n_samples_range.max() > 10_000:
        n_samples_scale = 1000
    else:
        n_samples_scale = 1
    
    for n_samples in n_samples_range:
        for crit in criteria:
            model = DecisionTreeRegressor(criterion=crit, **kwargs)
            bench.append(
                delayed(
                    model,
                    tags={"n_samples": n_samples / n_samples_scale, "criterion": crit},
                ).fit(X[:n_samples, :n_features], y[:n_samples])
            )
    res = Benchmark(repeat=repeat, **{metric_name: True})(bench)
    if repeat > 1:
        res = res.loc[:, (metric_name, "mean")].unstack(-1)
    else:
        res = res.loc[:, metric_name].unstack(-1)
    ax = res.plot(marker="o", ax=ax)
    ax.set(
        title=f"n_features={n_features}",
        xlim=(0, n_samples_range.max() / n_samples_scale),
    )
    if metric_name == "wall_time":
        ax.set(ylabel="wall time (s)")
    elif metric_name == "peak_memory":
        ax.set(ylabel="peak memory (MB)")
    if n_samples_scale > 1:
        ax.set(xlabel="n_samples (10³)")
    else:
        ax.set(xlabel="n_samples")
        
    return res, ax


fig, ax = plt.subplots(1, 2, figsize=(12, 4))

kwargs = {}

res_f10, _ = make_criterion_benchmark(n_features=10, ax=ax[0])
res_f100, _ = make_criterion_benchmark(n_features=100, ax=ax[1])

with sklearn.config_context(print_changed_only=True):
    fig.suptitle("DecitionTreeRegressor"
        + " benchmark on make_regression() data"
    )


res_f10['ratio'] = res_f10['poisson'] / res_f10['mse']
res_f100['ratio'] = res_f100['poisson'] / res_f100['mse']

fig, axes = plt.subplots(1, 2, figsize=(12, 4))
res_f10['ratio'].plot(ax=axes[0])
res_f100['ratio'].plot(ax=axes[1])

for ax in axes:
    ax.set(ylabel="wall time ratio poisson/mse")

fig.suptitle("DecitionTreeRegressor"
    " benchmark on make_regression() data")
```
</details>

lorentzenchr committed:
TST add further test for forbidden zero nodes

* Add further dataset to test_poisson_zero_nodes.

* This test fails as Poisson splitting tree predicts zero.

lorentzenchr commented:
I added a further dataset to `test_poisson_zero_nodes` where this test fails💣. This test checks that the poisson criterion can never have a node with value=0.
As the impurity improvement returns -INFINITY (same as initialized value), it should not allow those splits. I'm investigating the splitter code where this might happen.
Interestingly, the impurity value (~poisson loss) for those nodes that predict zero is `np.nan`.

lorentzenchr committed:
BUG check for y_mean <= 0 in poisson_loss

lorentzenchr committed:
CLN remove FIXME tag

lorentzenchr commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -1950,6 +1973,95 @@ def test_apply_path_readonly_all_trees(name):
+
+    # Choose a training set with non-negative targets (for poisson)
+    X, y = diabetes.data, diabetes.target
+    reg = Tree(criterion=criterion)
```
Good question! I can only tell that the test has to hold regardless of `random_state`. It is a property between tree and training data no matter how the tree parameters (except criterion) was set.

lorentzenchr committed:
CLN address review comments

thomasjpfan reviewed (approved):
LGTM

Please add an entry to the change log at `doc/whats_new/v0.24.rst` with tag |Feature|. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`.

NicolasHug commented on doc/modules/tree.rst:
```diff
@@ -429,99 +433,109 @@ Mathematical formulation
+
+Setting `criterion="poisson"` might be a good choice if your target is a count
+or a frequency (count per some unit). In any case, :math:`y >= 0` is a
+necessary condition to use this criterion.
```
I think we should document that it may be much slower than MSE.
Maybe we can make the same comment about MAE (it's much slower than MSE)

lorentzenchr commented on sklearn/tree/_criterion.pyx:
```diff
@@ -1325,3 +1325,122 @@ cdef class FriedmanMSE(MSE):
+
+                impurity += w * xlogy(self.y[i, k], self.y[i, k] / y_mean)
+
+        return impurity / (self.weighted_n_node_samples * self.n_outputs)
```
There is `check_sample_weights_invariance` in `estimator_checks.py`. IIUC, the common tests use only the default arguments, right? That means, only `DecisionTreeRegressor(criterion="mse"`) is tested.

Can I add `check_sample_weights_invariance` here in `test_tree.py` (for all regression criteria)?

lorentzenchr committed:
TST add tests for sample_weight consistency

lorentzenchr committed:
DOC add what_new entry

lorentzenchr committed:
DOC mention poisson and mae fit slower than mse

lorentzenchr commented:
I added tests for `sample_weight` consistency, a whats new entry and mentioned slower fitting time for poisson and mae. Please note these changes before merging. And thank you both for your reviews!

cmarmo commented:
@glemaitre, two approvals here: shall we merge? Thanks!

glemaitre commented on sklearn/tree/_criterion.pyx:
```diff
@@ -1330,3 +1330,122 @@ cdef class FriedmanMSE(MSE):
+            else:
+                y_mean_left = self.sum_left[k] / self.weighted_n_left
+                y_mean_right = self.sum_right[k] / self.weighted_n_right
+                proxy_impurity_left -= y_mean_left * log(y_mean_left)
```
Is it on purpose to not use `xlogy` here and use it below?

lorentzenchr replied:
Here, it can't happen that y_mean_... is zero. `xlogx = x * log(y)` is particularly useful in situations where `x=0` can happen with the expected overall result being zero.

glemaitre replied:
OK I see, cool.

glemaitre commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -208,18 +214,26 @@ def test_weighted_classification_toy():
+    if criterion == "poisson":
+        # make target positive
+        a = np.abs(np.min(y)) + 1
+        y_train = np.array(y) + a
```
So here `y` and `true_results` are a list and making an array will make a copy.
However, this is not super explicit.

Could we trigger an explicit copy first of `y` and `true_results` with a comment?

lorentzenchr replied:
I added a comment. Is that enough?

I think you already know, but note that `np.array(..., copy=True)` is already the default. On top of that, the `+ a` creates another new array again.

glemaitre replied:
Yes it is enough.

glemaitre commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -277,10 +291,11 @@ def test_diabetes_overfit(name, Tree, criterion):
-    "criterion, max_depth",
-    [("mse", 15), ("mae", 20), ("friedman_mse", 15)]
+    "criterion, max_depth, max_loss",
+    [("mse", 15, 60), ("mae", 20, 60), ("friedman_mse", 15, 60),
+     ("poisson", 15, 30)]
```
```suggestion
    [("mse", 15, 60),
     ("mae", 20, 60),
     ("friedman_mse", 15, 60),
     ("poisson", 15, 30)]
```

glemaitre commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -277,10 +291,11 @@ def test_diabetes_overfit(name, Tree, criterion):
 @pytest.mark.parametrize(
-    "criterion, max_depth",
-    [("mse", 15), ("mae", 20), ("friedman_mse", 15)]
+    "criterion, max_depth, max_loss",
```
I would be in favour to pass the metric in the parametrization as well.
It avoids the if/else condition within the test and will be explicit in the name while executing the test

glemaitre reviewed (approved):
LGTM. Only 2-3 nits.
I would be in favour of merging afterwords.

lorentzenchr committed:
TST add comment about not touching original data

lorentzenchr committed:
CLN nicer code formating

lorentzenchr committed:
TST parametrize metric as well

lorentzenchr commented:
@glemaitre Thanks for looking into this PR. I addressed your comments.

Note that this PR enables `criterion="poisson"` not only in `DecisionTreeRegressor` but also automatically in
- `tree.ExtraTreeRegressor`
- `ensemble.ExtraTreesRegressor`
- `ensemble.RandomForestRegressor`

Should I mention this in the what's new entry? So far, I did not as tests are not yet implemented for those.
So, I'd prefer a thorough separate PR after v0.24.

glemaitre commented:
> Should I mention this in the what's new entry and docstring? So far, I did not as tests are not yet implemented for those.
So, I'd prefer a thorough separate PR after v0.24.

I agree that I would prefer a separate PR. In addition to the tests and docstring, we might need to illustrate via an example maybe.
I think that we will be ready for the release before this subsequent PR.

glemaitre commented:
Oopsy I forgot to remove [MRG], my bad.
Thanks @lorentzenchr
